bulldozer/signals.go

```diff
@@ -21,109 +21,389 @@ import (
+
+// MatchesAny returns true if the pull request meets one or more signals. It also
 // returns a description of the signal that was met. The tag argument appears
 // in this description and indicates the behavior (trigger, ignore) this
 // set of signals is associated with.
-func (s *Signals) Matches(ctx context.Context, pullCtx pull.Context, tag string) (bool, string, error) {
+func (s *Signals) MatchesAny(ctx context.Context, pullCtx pull.Context, tag string) (bool, string, error) {
+	descriptions, matches, err := s.labelMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	if len(matches) > 0 {
+		return true, descriptions[0], nil
+	}
+
+	descriptions, matches, err = s.commentMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	if len(matches) > 0 {
+		return true, descriptions[0], nil
+	}
+
+	descriptions, matches, err = s.commentSubstringMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	if len(matches) > 0 {
+		return true, descriptions[0], nil
+	}
+
+	descriptions, matches, err = s.prBodyMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	if len(matches) > 0 {
+		return true, descriptions[0], nil
+	}
+
+	descriptions, matches, err = s.branchMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	if len(matches) > 0 {
+		return true, descriptions[0], nil
+	}
+
+	descriptions, matches, err = s.branchPatternMatches(ctx, pullCtx, tag)
+	if err != nil {
+		return false, "", err
+	}
+
+	if len(matches) > 0 {
+		return true, descriptions[0], nil
+	}
+
+	return false, fmt.Sprintf("pull request does not match the %s", tag), nil
+}
+
+// labelMatches determines which label signals match the given PR. It returns:
+// - A list of descriptions for each matched signal
+//   - These will only include the first item the signal matched when there is more than one
+// - A list of the matched signals
+func (s *Signals) labelMatches(ctx context.Context, pullCtx pull.Context, tag string) ([]string, Signal, error) {
 	logger := zerolog.Ctx(ctx)
+	matches := Signal{}
+	descriptions := []string{}
+
+	if len(s.Labels) == 0 {
+		logger.Debug().Msgf("No label singals have been provided to match against")
+		return descriptions, matches, nil
+	}
 
 	labels, err := pullCtx.Labels(ctx)
 	if err != nil {
-		return false, "unable to list pull request labels", err
+		return descriptions, matches, errors.Wrap(err, "unable to list pull request labels")
 	}
 
 	if len(labels) == 0 {
 		logger.Debug().Msgf("No labels found to match against")
+		return descriptions, matches, nil
 	}
+
 	for _, signalLabel := range s.Labels {
 		for _, label := range labels {
 			if strings.EqualFold(signalLabel, label) {
-				return true, fmt.Sprintf("pull request has a %s label: %q", tag, signalLabel), nil
+				matches = append(matches, signalLabel)
+				descriptions = append(descriptions, fmt.Sprintf("pull request has a %s label: %q", tag, signalLabel))
 			}
 		}
 	}
```

Comment:
The new `Signal` type is mostly used as way to distinguish string slices when a function returns more than one, but I think it's a missed opportunity to abstract some of this functionality. `Signal` could define a function (or functions) to determine if an input matches. It would probably make sense to define multiple types of `Signal` for the different matches (literal, case-insensitive, regexp), but multiple methods on a single type could work too.